tests/references/SDFqEstimator/gen_sdfq.sage

```diff
@@ -0,0 +1,101 @@
+load("tests/references/helpers/attack_cost.sage")
+load("tests/references/helpers/cost.sage")
+
+from math import comb as binomial, log2
+from itertools import chain
+
+
+def lee_brickell_correction(k: int) -> float:
+    """
+    Calculate the correction term for the Lee-Brickell procedure.
+
+    Args:
+        k: The dimension of the code.
+
+    Returns:
+        The calculated correction term.
+
+    Notes:
+        See SDFqAlgorithms/leebrickell.py line 98/99
+    """
+    return log2(k) * 2 - log2(binomial(k, 2))
+
+
+def gen_sdfq_lee_brickell(inputs: list[tuple]):
+    """
+    Generate expected complexities for Lee-Brickell SDFq problems.
+
+    Args:
+        inputs: A list of tuples, each containing (n, k, w, q) for a Lee-Brickell SDFq problem.
+
+    Returns:
+        A list of tuples, each containing the input parameters and the corresponding expected complexity.
+    """
+
+    def gen_single_case(input: tuple):
+        n, k, w, q = input
+        expected_complexity = (
+            log2(ISD_COST(n, k, w, q)) + log2(n) - lee_brickell_correction(k)
+        )
+        return input, expected_complexity
+
+    inputs_with_expected_outputs = list(map(gen_single_case, inputs))
+    return inputs_with_expected_outputs
+
+
+def gen_sdfq_stern(inputs: list[tuple]):
+    """
+    Generate expected complexities for Stern SDFq problems.
+
+    Args:
+        inputs: A list of tuples, each containing (n, k, w, q) for a Stern SDFq problem.
+
+    Returns:
+        A list of tuples, each containing the input parameters and the corresponding expected complexity.
+    """
+
+    def gen_single_case(input):
+        n, k, w, q = input
+        expected_complexity, _, _ = peters_isd(n, k, q, w)
+        return input, expected_complexity
+
+    inputs_with_expected_outputs = list(map(gen_single_case, inputs))
+    return inputs_with_expected_outputs
```

Comment:
Could this maybe be handled more elegantly by providing a generic function that produces the test cases once ``gen_single_case`` and the algorithm name is provided?

Reply:
So, something like this?

```python
def general_test(test_single_case, test_name, test_data):
    inputs_with_expected_outputs = test_data(test_name)
    map(test_single_case, inputs_with_expected_outputs)

# ...
def test_single_case_sdfq_lee_brickell(input_with_exp_output, epsilon):
    """
    Single test for the LeeBrickell-SDFq problem with an error tolerance of epsilon.
    """
    input, expected_complexity = input_with_exp_output
    n, k, w, q = input
    p = 2

    actual_complexity = LeeBrickell(
        SDFqProblem(n, k, w, q, **PARAMS), **PARAMS
    ).time_complexity(p=p)

    assert abs(actual_complexity - expected_complexity) < epsilon

general_test(test_single_case_sdfq_lee_brickell, "test_sdfq_lee_brickell", test_data)
```

Instead of this

```python

def test_sdfq_lee_brickell(test_data):
    """
    Test for the LeeBrickell-SDFq problem with an error tolerance of epsilon.
    """
    inputs_with_expected_outputs = test_data()
    epsilon = 0.01

    def test_single_case(input_with_exp_output):
        input, expected_complexity = input_with_exp_output
        n, k, w, q = input
        p = 2

        actual_complexity = LeeBrickell(
            SDFqProblem(n, k, w, q, **PARAMS), **PARAMS
        ).time_complexity(p=p)

        assert abs(actual_complexity - expected_complexity) < epsilon

    map(test_single_case, inputs_with_expected_outputs)
```

While that removes the code duplication, we now have some more troubles:

1. We have darkened the test process with another intermediary function.
2. We have splitted the information related to one test, into two different blocks (the one that defines the single_case_test, and the one that calls the evaluation of that `single_case_test`, via the `general_test_function`).
3. The argument `test_data` used in the *definition* of the functions is automatically provided by python fixtures when you execute the `pytest` command. I'm not sure that parameter can be "passed" into a function call (I can search if you want).

The biggest drawback to me is the number 2. We are saving nothing more than one line of code per test, because Python doesn't support multiline lambdas (to define the `test_single_case` directly in the `general_test` function call, for example), in exchange for a more convoluted test definition process.

Reply:
As I see it you could even automatize the call to

```general_test(...)```

similar to how we add the algorithm objects as attributes to an estimator object (in the estimator constructor).

All logic for the test the developer ever has to deal with would then be in the function ```test_sdfq_lee_brickell```.
Point 3 is not very clear to me yet.

Reply:
I'm in favor of keeping it as it is. We are talking about 100 lines of code. I think its fine if we loose a few lines here, but keep the whole thing readable, even for someone new to the project.

Reply:
@Memphisd 

> All logic for the test the developer ever has to deal with would then be in the function test_sdfq_lee_brickell.
Point 3 is not very clear to me yet.


1. The thing with pytest is... It collects only *function definitions* starting with the `test_` prefix, but it doesn't execute the file like a typical `python3 file.py` run does. That makes your approach based on *function calls* instead of *function definitions* hard to implement. And while I'm sure what you want is doable, it could involve a lot of work overwriting the collection method that pytest uses by default. Another option could be to create a big test, with each subtest as a inner definition function:

```python
def test_sdfq(test_data):
  def general_test(test_single_case, test_name, test_data):
    pass

  def sdfq_lee_brickell(input_with_exp_output):
     pass

  general_test(sdfq_lee_brickell, "sdfq_lee_brickell", test_data):
     pass
  # And repeat this with each test
```

While I could expect this to work, I still don't see the gain. We are obscuring the testing process by adding another intermediary function in exchange for 1 or 2 lines of code per-test.

2. The test developers *always* should have in mind that  they need to deal with three files, the `gen_estimator` file, the inputs dictionary file, and the `test_`file. They can't be understood as unrelated things, because each one represent a different step in the testing flow.

3. And in my previous last point, about pytest fixtures, what I want to say is: the parameter `test_data` I'm giving to each test, is provided automatically by a pytest feature called `fixture`. That fixture is configured in the `conftest.py` file, and it's because of this that we will load the reference `.yaml` only one time on each `pytest` command execution (even if we are running all the estimators tests). 
Because it is provided directly in the *function definition* and **not** in *function calls*, if we want to achieve something like what you are purposing, I should need to go really deep in the pytest module to understand how to capture that information and then pass it as an argument for the general testing function call.

While I have the same concern as you about the code duplication, what I found is that any alternative I can find or think of, involves either obscuring the testing process, or using advanced pytest testing features/hacky techniques. So I prioritize the maintanability of the code by using the most straightforward implementation I can, even if it involves those extra lines.

Reply:
> I'm in favor of keeping it as it is. We are talking about 100 lines of code. I think its fine if we loose a few lines here, but keep the whole thing readable, even for someone new to the project.

I think it is not only about code duplication here but about how to write the tests. I think the nested function structure overcomplicates it. All tests for all functions follow basically the same principle, its a KAT test with (in some cases) a certain tolerance. However the current framework requires you to write the KAT test logic, that is read the inputs and outputs create the objects, check against output with certain tolerance, for every test. I am questioning if we can do better here.
I think it is worth spending a few thoughts on it now. Once implemented, it will be much more work should we at some point decide to change it. However, if after careful consideration you still say there are too many technical obstacles in automatizing the logic, I guess we can proceed.

Reply:
@Dioprz is it clear what @Memphisd suggestion here is? If so, did you have time to think about a solution?

Reply:
No, not really. I can't find any better solution to what I proposed.

My line of thinking is like this:

1. We have some redundant lines of code, corresponding to the KAT tests logic.
2. If we want to remove them, we need to abstract those lines with either a class or a function.
3. A class is completely overkilled for this purposes.
4. A function that abstract those lines, must then be called with the non-redundant information as parameters, i.e.:
    1. The function that evaluates inputs to produce our estimators outputs.
    2. The deserialized dictionary.
    3. The function name, so we can access the correspondant data (inputs/exp_outputs) in the deserialized dictionary. (I think this could be abstracted too... Maybe I can infer it from the function name if they match... But anyways).
5. Because pytest ONLY collects test *definitions* starting by the `test_` prefix, the function call must be inside a bigger function starting with such prefix.
6. That bigger function must contain all the non-redudant information for each test, and then call the general function on that information.

This is what led me to propose the code I proposed in my previous response. And there I mention what drawbacks I see in such solution.

Maybe I'm making an incorrect deduction in some step, that can be solved in other, more efficient way, but I can't see it.

And, just for the record, I`m sure this lines could be abstracted in a neat way from a test-developer perspective by using advanced pytest features (like implementing our custom pytest collector function, or search if pytest could trigger the call for functions from a centralized file, for example). But this will take me a considerable amount of time to learn how to do it, and then implement it; not to mention that it will result weird to someone already familiar with pytest, as we are changing the game rules for the tool (pytest).

In case it helps in some way, the nested function structure can be removed. I like to think with maps and data transformations, but the logic can be translated into a for-loop for a more imperative approach if that sound better to you:

```python
def test_sdfq_lee_brickell(test_data):
    """
    Test for the LeeBrickell-SDFq problem with an error tolerance of epsilon.
    """
    inputs_with_expected_outputs = test_data()
    epsilon = 0.01

    def test_single_case(input_with_exp_output):
        input, expected_complexity = input_with_exp_output
        n, k, w, q = input
        p = 2

        actual_complexity = LeeBrickell(
            SDFqProblem(n, k, w, q, **PARAMS), **PARAMS
        ).time_complexity(p=p)

        assert abs(actual_complexity - expected_complexity) < epsilon

    map(test_single_case, inputs_with_expected_outputs)
```

Becomes

```python
def test_sdfq_lee_brickell(test_data):
    """
    Test for the LeeBrickell-SDFq problem with an error tolerance of epsilon.
    """
    inputs_with_expected_outputs = test_data()
    epsilon = 0.01

    for input, expected_complexity in inputs_with_expected_outputs:
        input, expected_complexity = input_with_exp_output
        n, k, w, q = input
        p = 2

        actual_complexity = LeeBrickell(
            SDFqProblem(n, k, w, q, **PARAMS), **PARAMS
        ).time_complexity(p=p)

        assert abs(actual_complexity - expected_complexity) < epsilon
```

Reply:
I think, we have something more like this in mind:
```python
# NOTE: the missing `test_` prefix, s.t. the function is ignored by the pytest framework
# This would be in `tests/validations/test_sdfq.py`
def single_case_leebrickell(input): 
    n, k, w, q = input 
    return LeeBrickell(SDFqProblem(n, k, w, q)).time_complexity(p=2)

# This would be in `tests/validations/test_sdfq.py`
def single_case_stern(input): 
    n, k, w, q = input 
    return Stern(SDFqProblem(n, k, w, q)).time_complexity()

# NOTE: this function would still be in conftest.py
def test_data(yaml_references: dict):
    """
    Fixture to retrieve test data from YAML references.
    """

    def get_test_data(test_file: str = None, test_function: str = None):
        """
        Get inputs and expected outputs for a specific test function.
        If test_file or test_function is not provided, it will be inferred from the caller.
        """
        if test_file is None or test_function is None:
            caller_frame = inspect.currentframe().f_back
            if test_file is None:
                test_file = os.path.splitext(
                    os.path.basename(caller_frame.f_code.co_filename)
                )[0]
            if test_function is None:
                test_function = caller_frame.f_code.co_name

        expected_complexity, input_data, epsilon = yaml_references[test_file][test_function]
        # admittedly: this is shady, as you need catch the case that multiple test cases are
        # mapped to the same algorithm, like `gen_stern` and `gen_stern_range` in our sdfq example.
        actual_complexity = test_function(*input_data)
        assert abs(actual_complexity - expected_complexity) < epsilon
```
The idea is to reduce the usage of `pytest` to basically a single function: `test_data()`, which does everything. From the test-case extraction to the actual complexity comparison. 

This has the big advantage that the work a new contributor would have to do, is minimal. Basically write a function for each algorithm the new estimator contains (in this example only `single_case_leebrickell` and `single_case_stern`)

@Dioprz  please let me know whether this is technically feasible at all. Specially this mapping of multiple testcase to a single test function (in our example sdfq_stern).